modules/gapi/src/streaming/onevpl/accelerators/surface/surface_pool.cpp

```diff
@@ -0,0 +1,78 @@
+#include "streaming/onevpl/accelerators/surface/surface_pool.hpp"
+#include "streaming/onevpl/accelerators/surface/surface.hpp"
+#include "logger.hpp"
+
+#ifdef HAVE_ONEVPL
+
+namespace cv {
+namespace gapi {
+namespace wip {
+
+void CachedPool::reserve(size_t size) {
+    surfaces.reserve(size);
+}
+
+size_t CachedPool::size() const {
+    return surfaces.size();
+}
+
+void CachedPool::clear() {
+    surfaces.clear();
+    next_free_it = surfaces.begin();
+    cache.clear();
+}
+
+void CachedPool::push_back(surface_ptr_t &&surf) {
+    cache.insert(std::make_pair(surf->get_handle(), surf));
+    surfaces.push_back(std::move(surf));
+    next_free_it = surfaces.begin();
+}
+
+CachedPool::surface_ptr_t CachedPool::find_free() {
```

Comment:
Is this pool supposed to be thread safe or the protection is done outside?

Reply:
MediaFrames may be destroyed concurrently, is it ok for this implementation?

Reply:
yes, it is ok. It just decreased surface shared counter as `shared_ptr<surface>` and ApplicationCounter inside `surface`